engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/base/DefaultTextExtractor.java

```diff
@@ -390,13 +390,35 @@ protected void processParagraphChilds( final ParagraphRenderBox box ) {
         if ( lineBox.getStaticBoxLayoutProperties().isPreserveSpace() == false ) {
           addSoftBreak();
         } else {
-          addEmptyBreak();
+
+          RenderNode deepestLastChild = getDeepestLastChild( lineBox );
+          if ( deepestLastChild != null && deepestLastChild instanceof RenderableText ) {
+            RenderableText renderableText = (RenderableText) deepestLastChild;
+            int wordSize = renderableText.getGlyphs().getSize();
+            int currentOffset = renderableText.getOffset();
+            int currentLength = renderableText.getLength();
+
+            if ( wordSize == currentOffset + currentLength ) {
+              addSoftBreak();
+            } else {
+              addEmptyBreak();
+            }
+          } else {
+            addSoftBreak();
+          }
         }
       }
       lineBox = (RenderBox) lineBox.getNext();
     }
   }
 
+  protected RenderNode getDeepestLastChild( RenderBox lineBox ) {
+    if ( lineBox.getLastChild() != null && lineBox.getLastChild() instanceof RenderBox ) {
+      return getDeepestLastChild( lineBox );
```

Comment:
Infinite loop? You probably wanted to pass a cast version of lineBox.getlastChild instead.

Reply:
Ups, that's right!
PR updated.